src/rvsdg/egglog_optimizer/ivt.egg

```diff
@@ -0,0 +1,115 @@
+       (let outer-else-block (BaseChangeVecOperand outputs gamma inner-else-block))
+       (let inner-then-block (OperandGroup (SubstVecOperandAll (vec-get gamma-out 1) gamma-in)))
+       (let inner-theta (Theta (BaseChangeOperand pred gamma inner-then-block) 
+                               (PassThroughArguments (vec-length inputs))
+                               (BaseChangeVecOperand outputs gamma inner-then-block)))
+        (union loop 
+           (Gamma (SubstOperandAll pred (VO inputs)) 
+                  (VO inputs) 
+                  (VVO 
+                   (vec-of 
+                       outer-else-block
+                       (BodyToVecOperand (vec-length inputs) inner-theta)))))))
+
+(rewrite (BaseChangeExpr (Const ty ops lit) from to)
+  (Const ty ops lit) :ruleset subst)
+(rewrite (BaseChangeExpr (Call ty f args n-outs) from to)
+  (Call ty f (BaseChangeVecOperand args from to) n-outs) :ruleset subst)
+(rewrite (BaseChangeExpr (badd ty a b) from to) 
+  (badd ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bsub ty a b) from to) 
+  (bsub ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bmul ty a b) from to) 
+  (bmul ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bdiv ty a b) from to) 
+  (bdiv ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (beq ty a b) from to)
+  (beq ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (blt ty a b) from to)
+  (blt ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bgt ty a b) from to)
+  (bgt ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (ble ty a b) from to)
+  (ble ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bge ty a b) from to)
+  (bge ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bnot ty a) from to)
+  (bnot ty (BaseChangeOperand a from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (band ty a b) from to)
+ (band ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (bor ty a b) from to)
+  (bor ty (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+(rewrite (BaseChangeExpr (PRINT a b) from to)
+  (PRINT (BaseChangeOperand a from to) (BaseChangeOperand b from to)) :ruleset subst)
+
+(rewrite (BaseChangeOperand (Arg x) from to) (Arg x) :ruleset subst)
+(rewrite (BaseChangeOperand (Node from) from to) (Node to) :ruleset subst)
+(rewrite (BaseChangeOperand (Project i from) from to) (Project i to) :ruleset subst)
```

Comment:
What about the case where we have `(BaseChangeOperand (Project i child) from to)` where `child != from`?
Are you going to add a rule for this case or is it not needed?

If you add a rule for that case, it may be unsound (consider what happens if child then later becomes equal to from).

Reply:
Right.. I guess I was thinking a similar reasoning applies here compared with other `Subst` rules. We can have separate rules for `(subst i v (Arg i))` and `(subst i v (Arg j)) :where (!= i j)` because we know `(Arg i)` and `(Arg j)` will never be equated later.

But I see your point: I don't think that actually holds for the `Body`s that can show up here. A lot of rules merge two Body's directly. I'll think about this a bit. Let me know if you have something in mind.

Reply:
Ryan's can-subst-beneath rules are made for this same situation! They are clever- they substitute bottom up and union at the nearest context.